Value-Based Classes

nicolaiparlog opened the pull request:
:bangbang: **THIS PULL REQUEST IS NOT COMPLETE YET**

This is my shot at implementing [Feature Request #313: Warn about identity-sensitive operations on value based classes ](https://sourceforge.net/p/findbugs/feature-requests/313/).

For brevity's sake I'll use _vbi_ for _value-based instance(s)_, i.e. an instance of a value-based class.
## Current State

The annotation `@ValueBased` can be used to mark own value-based classes. These classes and those from the JDK are processed uniformly when it comes to the following checks:
- :white_check_mark: no reference comparison
- :white_check_mark: serialization of _vbi_
  - :white_check_mark: no non-transient field in serializable class (works out of the box)
  - :white_check_mark: no direct serialization via `ObjectOutputStream.writeObject`  (works out of the box)
- :red_circle: identity hash code
  - :white_check_mark: no argument to `System.identityHashCode`
  - :red_circle: no key in an `IdentityHashMap`
- :white_check_mark: no locking on a _vbi_
  - :white_check_mark: no use in `synchronized` block
  - :white_check_mark: no calls to `Object.wait`
  - :white_check_mark: no calls to `Object.notify`
  - :white_check_mark: no calls to `Object.notifyAll`
## Notes

Ensuring that a class annotated with `@ValueBased` has all properties described in the [official documentation](https://docs.oracle.com/javase/8/docs/api/java/lang/doc-files/ValueBased.html) is not in the scope of [#313](https://sourceforge.net/p/findbugs/feature-requests/313/) and hence also not of this pull request.

amaembo commented:
Re: [findbugs] Value-Based Classes (#46)


Hello, Nicolai!

Yes, I'm remember about your pull-request. It's actually very big feature, and I have some doubts about some particular sub-features, thus I cannot just agree and merge. It's too much responsibility for me as for the newest member of FindBugs team.
I will try to review your code and probably add some comments. Meanwhile I would like to ask Andrei and Bill. What do you think in general? If you don't have the time to dig into the details, at least say whether in general you like these new features or not. See
http://sourceforge.net/p/findbugs/feature-requests/313/ for details.

With best regards,
Tagir Valeev.





 THIS PULL REQUEST IS NOT COMPLETE YET
This is my shot at implementing Feature Request #313: Warn about identity-sensitive operations on value based classes .
For brevity's sake I'll use vbi for value-based instance(s), i.e. an instance of a value-based class.
Current State
The annotation @ValueBased can be used to mark own value-based classes. These classes and those from the JDK are processed uniformly when it comes to the following checks:


 no reference comparison
 serialization of vbi 


 no non-transient field in serializable class (works out of the box)
 no direct serialization via ObjectOutputStream.writeObject (works out of the box)
 identity hash code 


 no argument to System.identityHashCode
 no key in an IdentityHashMap
 no locking on a vbi 


 no use in synchronized block
 no calls to Object.wait
 no calls to Object.notify
 no calls to Object.notifyAll

Notes
Ensuring that a class annotated with @ValueBased has all properties described in the official documentation is not in the scope of #313 and hence also not of this pull request.
You can view, comment on, or merge this pull request online at:
  https://github.com/findbugsproject/findbugs/pull/46
Commit Summary


Create a test for a non-transient field of a value-based type in a serializable class.
Identify the situation when an instance of a value based type is compared using reference equality.
Identify value-based classes with an annotation
Move identification of value-based classes into own utility class
Add tests for writing value-based instances to 'ObjectOutputStream.writeObject'
Identify all value-based classes
Remove undesired Warning from test case
Name class in singular
Create a detector for locking on value-based classes
Make contract of 'ValueBasedClassIdentifier.isValueBasedClass(String)' more precise
Create detector for all 'Object.wait' method calls on value-based classes
Move VBC tests into one class 'sfBugsNew/Feature313jadk8'
Use slashed class names and include patterns for notification in vbi
Fix use of "VBC_MO_*" bug pattern names
Implement pattern for calling 'System.identityHashCode' with a value-based class

File Changes


M findbugs/etc/findbugs.xml (7) 
M findbugs/etc/messages.xml (14) 
A findbugs/src/java/edu/umd/cs/findbugs/annotations/ValueBased.java (80) 
A findbugs/src/java/edu/umd/cs/findbugs/ba/vbc/ValueBasedClassIdentifier.java (121) 
M findbugs/src/java/edu/umd/cs/findbugs/detect/FindRefComparison.java (61) 
A findbugs/src/java/edu/umd/cs/findbugs/detect/IdentityHashCodeOnValueBasedClass.java (66) 
A findbugs/src/java/edu/umd/cs/findbugs/detect/MonitoringValueBasedClass.java (107) 
A findbugsTestCases/src/java/sfBugsNew/Feature313jdk8.java (264) 

Patch Links:


https://github.com/findbugsproject/findbugs/pull/46.patch
https://github.com/findbugsproject/findbugs/pull/46.diff

—
Reply to this email directly or view it on GitHub.

nicolaiparlog commented:
> It's actually very big feature, and I have some doubts about some particular sub-features, thus I cannot just agree and merge. It's too much responsibility for me as for the newest member of FindBugs team.

No worries, I'm expecting a thorough code review, so I'm up for all kinds of critical comments and change requests.

iloveeclipse commented:
Hi,

I briefly looked over the requirements & feature description, it sounds interesting, so I will not be against including this PR as soon as the review is complete.

Just wondering if anyone from JDK team could confirm that the "value types" feature is still planned and worked on - the last date of the document [1] is April 2014.

[1]	http://cr.openjdk.java.net/~jrose/values/values-0.html


On Sunday 30 August 2015 05:06 Tagir Valeev wrote:
> Re: [findbugs] Value-Based Classes (#46)
> 
> 
> Hello, Nicolai!
> 
> Yes, I'm remember about your pull-request. It's actually very big feature, and I have some doubts about some particular sub-features, thus I cannot just agree and merge. It's too much responsibility for me as for the newest member of FindBugs team.
> I will try to review your code and probably add some comments. Meanwhile I would like to ask Andrei and Bill. What do you think in general? If you don't have the time to dig into the details, at least say whether in general you like these new features or not. See
> http://sourceforge.net/p/findbugs/feature-requests/313/ for details.
> 
> With best regards,
> Tagir Valeev.


-- 
Kind regards,
google.com/+AndreyLoskutov

nicolaiparlog commented:
In August 2015 Brian Goetz discussed [Adventures on the Road to Valhalla](https://www.youtube.com/watch?v=uNgAFSUXuwc) at JVMLS (see my writeup of it [here](http://blog.codefx.org/java/dev/the-road-to-valhalla/)). While it focuses on specialization (i.e. enabling `ArrayList<int>`), this topic is brought forward by value types which [he explains in the prologue](https://www.youtube.com/watch?v=uNgAFSUXuwc#t=2m24s).

On the Valhalla mailing list Oracle's [Gavin Bierman posted this on August 3rd](http://mail.openjdk.java.net/pipermail/valhalla-dev/2015-August/001256.html):

> In Valhalla we will be in the world where some type variables will range over all reference types, and some over other sorts. It’s this 'other sorts' stuff that is new. In our experiments, we have identified three useful sorts: (i) `any` the sort containing all types, (ii) `ref` the sort containing all reference types (currently, the only sort denotable in Java), and (iii) `val` the sort containing all value types (currently the eight primitive types, but we are also experimenting with supporting user-defined value classes).

That being said, this is of course all experimental and it is by no means guaranteed that

1. value types will indeed come to Java
2. the limitations on value-based classes are still relevant then

amaembo commented on findbugs/etc/findbugs.xml:
```diff
@@ -621,6 +621,13 @@
+                    reports="VBC_IDENTITY_HASHCODE"/>
+                    
+          <BugCode abbrev="VBC_MO_LOCK"/>
+          <BugCode abbrev="VBC_MO_WAIT"/>
```
I see no need to declare separate abbreviations VBC_MO_LOCK, etc. Instead you should declare the bug patterns like:

    <BugPattern abbrev="VBC" type="VBC_MO_LOCK" category="BAD_PRACTICE"/>

And the same row should be added for other newly-defined patterns (`VBC_MO_WAIT`, `VBC_MO_NOTIFY`, `VBC_IDENTITY_HASHCODE`). The `abbrev` is somewhat a group of related bugs. I think, `VBC` is good for all new bug patterns you introduce.

amaembo commented on findbugs/etc/messages.xml:
```diff
@@ -1620,6 +1620,20 @@ factory pattern to create these objects.</p>
+  <Detector class="edu.umd.cs.findbugs.detect.IdentityHashCodeOnValueBasedClass">
+    <Details>
+      <![CDATA[
+        <p>TODO</p>
```
You should fill the detector information here and (more important) create information sections about separate bug patterns (`<BugPattern type="VBC_MO_WAIT">`, etc.) and the information about `VBC` group (`<BugCode abbrev="VBC">Value-based classes</BugCode>`). Put them into the separate groups.

amaembo commented on findbugs/src/java/edu/umd/cs/findbugs/ba/vbc/ValueBasedClassIdentifier.java:
```diff
@@ -0,0 +1,121 @@
+        try {
+            ClassDescriptor classDescriptor = DescriptorFactory.createClassDescriptor(className);
+            XClass xClass = Global.getAnalysisCache().getClassAnalysis(XClass.class, classDescriptor);
+            ClassDescriptor annotationDescriptor = DescriptorFactory.createClassDescriptor(ValueBased.class);
```
`ClassDescriptor annotationDescriptor = DescriptorFactory.createClassDescriptor(ValueBased.class);` can be stored in static field for better performance.

nicolaiparlog commented on findbugs/src/java/edu/umd/cs/findbugs/ba/vbc/ValueBasedClassIdentifier.java:
```diff
@@ -0,0 +1,124 @@
+            AnnotationValue annotation = xClass.getAnnotation(VALUE_BASED_ANNOTATION_DESCRIPTOR);
+            return annotation != null;
+        } catch (CheckedAnalysisException ex) {
+            // TODO (nipa@codefx.org) log the exception?
```
What should I do here? Some call sites (e.g. `FindRefComparison.checkRefComparison`) were ok if this method were to throw the exception; others (e.g. `MonitoringValueBasedClass.sawOpcode`) not.

amaembo replied:
You can use `AnalysysContext.logError(message, ex)` and return false.

nicolaiparlog commented on findbugs/src/java/edu/umd/cs/findbugs/detect/IdentityHashCodeOnValueBasedClass.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public void sawOpcode(int seen) {
+        // TODO (nipa@codefx.org) define these patterns and determine priority
```
What priority would you recommend for this bug?

amaembo replied:
Use NORMAL priority.

nicolaiparlog commented on findbugs/src/java/edu/umd/cs/findbugs/detect/MonitoringValueBasedClass.java:
```diff
@@ -0,0 +1,110 @@
+
+    @Override
+    public void sawOpcode(int seen) {
+        // TODO (nipa@codefx.org) define these patterns and determine priority
```
What priorities would you recommend for these bugs?

amaembo replied:
Use NORMAL priority

nicolaiparlog commented:
I addressed all comments and started clearing up my TODOs.

What remains is the check for not using value-based instances as keys in an `IdentityHashMap`. It would be great if someone pointed me to an example, where something similar is done.

amaembo commented on findbugs/src/java/edu/umd/cs/findbugs/detect/IdentityHashCodeOnValueBasedClass.java:
```diff
@@ -0,0 +1,69 @@
+    public IdentityHashCodeOnValueBasedClass(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
```
You may quickly bypass uninteresting classes which will make this detector faster.

1. Create a constant method descriptor:

        private static final MethodDescriptor IDENTITY_HASH_CODE = 
                new MethodDescriptor("java/lang/System", "identityHashCode", "(Ljava/lang/Object;)I", true);

2. Add the following check:

        @Override
        public void visitClassContext(ClassContext classContext) {
            if(hasInterestingMethod(classContext.getJavaClass().getConstantPool(), 
                                                       Collections.singleton(IDENTITY_HASH_CODE))) {
                super.visitClassContext(classContext);
            }
        }

So if class constant pool has no identityHashCode method, you skip the whole class which would be much faster than examining its code.

nicolaiparlog replied:
At first I though `hasInterestingMethod` was a method existing somewhere. I didn't find it, so [I implemented it myself](https://github.com/CodeFX-org/findbugs/commit/23e476959bb2b3b58f1055e2163ea1eb00d8376f#diff-4777956d7399da5348db12b0d1b22a98R61).

amaembo replied:
This method actually exists in PreorderVisitor which is in superclass hierarchy, so you can just call it. No reason to implement it by yourself.

nicolaiparlog replied:
Done in 8855342.

Apparently I didn't update my branch for almost a year :open_mouth: . Did that now and rebased my changes on top, so the commits listed below are largely the known ones from before - the first new one starts at 4b05469.

amaembo commented on findbugs/src/java/edu/umd/cs/findbugs/detect/IdentityHashCodeOnValueBasedClass.java:
```diff
@@ -0,0 +1,69 @@
+        boolean callToSystemIdentityHashCode = seen == INVOKESTATIC
+                && "java/lang/System".equals(getClassConstantOperand())
+                && "identityHashCode".equals(getNameConstantOperand());
+        if (!callToSystemIdentityHashCode) {
```
Also having `IDENTITY_HASH_CODE` constant, you can simplify this:

    boolean callToSystemIdentityHashCode = seen == INVOKESTATIC && getMethodDescriptorOperand().equals(IDENTITY_HASH_CODE);

amaembo commented on findbugs/src/java/edu/umd/cs/findbugs/detect/FindRefComparison.java:
```diff
@@ -1072,6 +1090,34 @@ private void handleSuspiciousRefComparison(JavaClass jclass, Method method, Meth
+            xf = ((FinalConstant) rhsType).getXField();
+        }
+        String sourceFile = jclass.getSourceFileName();
+        String bugPattern = "VBC_REF_COMPARISON";
```
The FindRefComparison detector must be marked in findbugs.xml as reporting "VBC_REF_COMPARISON" bug pattern

nicolaiparlog replied:
[Done](https://github.com/findbugsproject/findbugs/pull/46/files#diff-61b90f4ada57d0056fcacf3895ec6b6eL476).

amaembo commented on findbugs/src/java/edu/umd/cs/findbugs/detect/MonitoringValueBasedClass.java:
```diff
@@ -0,0 +1,109 @@
+        // check the argument to 'MONITORENTER', which is the last item on the stack
+        Item lastStackItem = getStack().getStackItem(0);
+        String classSignature = lastStackItem.getSignature();
+        String slashedClassName = ClassName.extractClassName(classSignature);
```
Seems that extractClassName is too smart for this case. If the type is the array type (like "[Ljava/time/LocalDate;"), it also extracts the original type. Here's the test case:

        @NoWarning("VBC_MO_LOCK")
        public void lockOnLocalDateArray(LocalDate[] lock) {
            synchronized (lock) {
                System.out.println("Locking on LocalDate.");
            }
        }

An array of value-based-classes is not value-based. While it seems strange to synchronize on array, it can be perfectly safe if array object is fully controlled by this code. In any case you should not report VBC_MO_LOCK here (the same for all other warnings). Add tests with arrays.

Probably better to extract type using ClassName.fromFieldSignature which extracts the className only if it's actually class reference and returns null in other cases (you should check for nulls)

amaembo commented on findbugs/src/java/edu/umd/cs/findbugs/detect/FindRefComparison.java:
```diff
@@ -967,16 +969,32 @@ private void checkRefComparison(Location location, JavaClass jclass, Method meth
+
+            String rhsSignature = rhsType.getSignature();
+            @SlashedClassName
+            String rhsSlashed = ClassName.extractClassName(rhsSignature);
```
Currently your code reports missing classes like "C", "Z", "I". I suspect that's because extractClassName leaves the argument as is if the class is primitive type and you try to load it in "isAnnotatedAsValueBasedClass". Try to use ClassName.fromFieldSignature(..) (you will probably need to check the result for null). Running ant in findbugsTestCases should not report any errors, nor any missing classes.

nicolaiparlog replied:
I hope the errors you report are gone now.

> Running ant in findbugsTestCases should not report any errors, nor any missing classes.

Running `ant` causes all kinds of errors and I must admit that I am hesitant to invest time into solving them - getting to a state where I can work with FindBugs was a little tedious and I am happy that it works at all.

amaembo commented on findbugs/src/java/edu/umd/cs/findbugs/ba/vbc/ValueBasedClassIdentifier.java:
```diff
@@ -0,0 +1,126 @@
+    private static boolean isAnnotatedAsValueBasedClass(@SlashedClassName String className) {
+        try {
+            ClassDescriptor classDescriptor = DescriptorFactory.createClassDescriptor(className);
+            XClass xClass = Global.getAnalysisCache().getClassAnalysis(XClass.class, classDescriptor);
```
(TODO reminder for myself) Have to check whether this slows down the analysis. This method is executed quite often, but 99.99% of classes are not annotated as VBC. It's possible that non-cached XClass objects are often requested leading to excessive class reloading.

nicolaiparlog replied:
I am curious what you'll find here. Be sure to let me know!

nicolaiparlog commented on findbugsTestCases/src/java/sfBugsNew/Feature313jdk8.java:
```diff
@@ -0,0 +1,314 @@
+            System.out.println("Reference comparison of 'Optional' yields: " + same);
+        }
+
+        // it would be great if this worked, but 'optionals.get(0)' gets recognized as string
```
Any ideas whether we can get this to work?

(There is a similar case [further below](https://github.com/findbugsproject/findbugs/pull/46/files#diff-5ced2e172ef0ce58b2a1ffd689b60e7aR261).)

nicolaiparlog commented:
Hey @amaembo , nice collaboration on this PR. I hope you're not running out of patience. :wink:

nicolaiparlog commented:
@amaembo I know you are doing this in your free time so I'm in no position to rush you. Still, it would be great if we could get this merged in the next couple of weeks. A bug workload is heading my way and I would like to get this done before it arrives.

nicolaiparlog commented:
There is no progress here because I don't know how to implement the check for not using value-based instances as keys in an `IdentityHashMap`. It would be great if someone pointed me to an example, where something similar is done.

iloveeclipse commented:
If you are still interested, please rebase your PR on top of SpotBugs and push it to https://github.com/spotbugs/spotbugs/.
